### ISSUE TITLE
[feature request] Add a localization sample

### ISSUE THREAD

**Issue by ozasadnyy** [Polymer/pwa-starter-kit#91]
It would be nice to have an example with the support of several languages.

**Comment by JosefJezek** [comment]
I found LocalizedLitElement https://github.com/Dabolus/localized-lit-element

**Comment by abdonrd** [comment]
@keanulee @notwaldorf @frankiefu (code owners) any recommendation?

**Comment by keanulee** [comment]
My gut feeling is that this is out of scope for pwa-starter-kit (boilerplate we don't want to maintain), but I do support the idea of subclassing the base element class to add specific features.

**Comment by abdonrd** [comment]
@keanulee maybe scoped from [pwa-helpers](https://github.com/Polymer/pwa-helpers)? Or directly discarded?

**Comment by abdonrd** [comment]
I did a quick test with [`i18next`](https://github.com/i18next/i18next).

#### `src/localize-mixin.js`

```javascript
let _i18nextInitialized = false;

export const localize = i18next => baseElement => class extends baseElement {
  _shouldRender(props, changedProps, old) {
    // Also check active property used by PageViewElement
    return changedProps && changedProps.active
      ? props.active && _i18nextInitialized
      : _i18nextInitialized;
  }

  connectedCallback() {
    if (!_i18nextInitialized) {
      i18next.on('initialized', options => {
        _i18nextInitialized = true;
        this.requestRender();
      });
    }

    i18next.on('languageChanged', () => {
      this.requestRender();
    });

    if (super.connectedCallback) {
      super.connectedCallback();
    }
  }
};
```

#### `src/i18next.js`

```javascript
import { use } from 'i18next/dist/es/index.js';
import Backend from 'i18next-xhr-backend/dist/es/index.js';

export const i18next = use(Backend).init({
  fallbackLng: 'en',
  debug: true,
  ns: ['app'],
  defaultNS: 'app',
  backend: {
    loadPath: 'locales/{{lng}}/{{ns}}.json',
  },
});
```

#### `src/components/my-view1.js`

```javascript
import { html } from '@polymer/lit-element';
import { PageViewElement } from './page-view-element.js';

import { localize } from '../localize-mixin.js';
import { i18next } from '../i18next.js';

// These are the shared styles needed by this element.
import { SharedStyles } from './shared-styles.js';

class MyView1 extends localize(i18next)(PageViewElement) {
  _render(props) {
    return html`
      ${SharedStyles}
      <section>
        <span on-click="${_ => i18next.changeLanguage('ja')}">JA</span>
        <h2>${i18next.t('title')}</h2>
        <p>This is a text-only page.</p>
        <p>It doesn't do anything other than display some static text.</p>
      </section>
      <section>
        <h2>Welcome</h2>
        <p>Lorem ipsum...</p>
      </section>
      <section>
        <p>Vestibulum at...</p>
      </section>
    `;
  }
}

window.customElements.define('my-view1', MyView1);
```

Thoughts?

**Comment by notwaldorf** [comment]
I've actually been looking at i18next this week as well. I'm not sure whether this would be part of a template, or just a feature added to one of the sample apps. I have something similar to what Abdon has, but I haven't figured out what app this should go in, or a compelling example.

I'm on vacation next week, so stay tuned. In the meantime, if anyone needs this urgently, by all means use Abdon's code! :)

**Comment by Dabolus** [comment]
I personally think that adding i18n to the main template would overcomplicate it for people who want to build their PWA using a single locale. What about adding a different template that features i18n? (e.g. `template-i18n`).

Another alternative would be to add i18n on the main template and create a `template-no-i18n`, like you did with `template-no-redux`

**Comment by notwaldorf** [comment]
@Dabolus I literally just said that -- that we're considering a different template, or a different app :)

**Comment by Dabolus** [comment]
@notwaldorf I'm sorry, I misinterpreted your comment 😅

**Comment by notwaldorf** [comment]
I have, but I don't have the bandwidth to take this over the finish line right now (trying to finish up the 3.0 polymer elements). Last I talked to @frankiefu and @keanulee, we were leaning towards maybe adding it to one of the apps, rather than a separate template. Can't speak for everyone else, but I would review a Flash Cards PR for example that added i18next

**Comment by evayde** [comment]
@abdonrd 
I would suggest adding annotations to your mixin, to avoid linter-warnings when building. 

```
/**
 * Localize mixin
 * 
 * @polymer
 * @mixinFunction
 **/
export const localize = i18next => baseElement => class extends baseElement {
// ...
```

Besides that awesome work.

**Comment by evayde** [comment]
@abdonrd 
Your mixin based on the new Lit-Element (lit-element@dev), lgtm

```
/**
 * Localize mixin
 * 
 * @polymer
 * @mixinFunction
 **/
export const localize = i18next => baseElement => class extends baseElement {
  shouldUpdate(changedProps) {
    // Also check active property used by PageViewElement
    return changedProps.has('active')
      ? changedProps.get('active') && _i18nextInitialized
      : _i18nextInitialized;
  }

  connectedCallback() {
    if (!_i18nextInitialized) {
      i18next.on('initialized', options => {
        _i18nextInitialized = true;
        this.invalidate();
      });
    }

    i18next.on('languageChanged', () => {
      this.invalidate();
    });

    if (super.connectedCallback) {
      super.connectedCallback();
    }
  }
};
```

**Comment by steamonimo** [comment]
@evayde Could you please provide a very simplified but complete example of how internationalization could and should be implemented with lit-element? I am new to PWA and these fragments for i18n provided by you and @abdonrd do not allow me to grasp the concept of i18n yet.

**Comment by evayde** [comment]
@steamonimo We depend on i18next https://www.npmjs.com/package/i18next
Also my implementation is only for the current dev of Lit-Element, not for the master branch. 

You could pull the pwa-starter-kit, add the files references by @abdonrd to your project and also provide your translationfiles, this would already be the minimal example. As you can see by this line `loadPath: 'locales/{{lng}}/{{ns}}.json'` we chose the locales folder to store language-files. The lng and ns attributed depend on your implementation if i18next. You can read more in their documentation how translations are working and how to extend their framework.

https://www.i18next.com/overview/introduction

If you have more questions please specify your exact problem.

**Comment by abdonrd** [comment]
@evayde the example (https://github.com/Polymer/pwa-starter-kit/issues/91#issuecomment-416022694) doesn't work for me:

```
Uncaught TypeError: this.invalidate is not a function
```

**Comment by keanulee** [comment]
With lit-element@dev, it will be `requestUpdate()`

**Comment by jorgecasar** [comment]
I agree with @abdonrd about #issuecomment-400424788 but I would like to include a detail about nomenclature. According with [W3C Localization vs. Internationalization](https://www.w3.org/International/questions/qa-i18n.en) I guess we are talking about localization instead of internationalization.

@notwaldorf @abdonrd do you agree to move this issue to [pwa-helpers](https://github.com/Polymer/pwa-helpers)?

**Comment by abdonrd** [comment]
@jorgecasar yes! I would be happy that something like `i18next-localize-mixin.js` was in the `pwa-helpers`!

**Comment by fleckdalm** [comment]
Any update on this?

**Comment by JosefJezek** [comment]
Here is mixin using i18next https://github.com/hamedasemi/lit-element-i18n
